clippy_lints/src/outer_doc_comment_bang.rs

```diff
@@ -0,0 +1,86 @@
+use clippy_utils::diagnostics::{multispan_sugg, span_lint_and_then};
+use if_chain::if_chain;
+use rustc_ast::{ast::*, token::CommentKind};
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects the use of outer doc comments (`///`, `/**`) followed by a bang (`!`).
+    ///
+    /// ### Why is this bad?
+    /// Triple-slash comments (known as "outer doc comments") apply to items that follow it.
+    /// An outer doc comment followed by a bang (i.e. `///!`) has no specific meaning.
+    ///
+    /// The user most likely meant to write an inner doc comment (`//!`, `/*!`), which
+    /// applies to the parent item (i.e. the item that the comment is contained in,
+    /// usually a module or crate).
+    ///
+    /// ### Known problems
+    /// The suggestion made by this lint may be invalid if it sees a normal outer doc comment (`///`)
+    /// followed by `///!`. This lint suggests changing it to `//!`, but that in turn makes the outer doc comment invalid.
```

Comment:
What I meant by "the suggestion may be invalid" is that the suggestion might not be valid code (i.e. does not compile). For example, writing
```rs
///
///!
fn foo() {}
```
compiles as is but triggers this lint and the lint suggests changing the last line to `//!`. Following the lint and changing it to 
```rs
///
//!
fn foo() {}
```
results in a compile error (so the suggestion is invalid).

Reply:
Ah, ok. Sorry, I mixed up stuff and misunderstood before.

Then I guess not linting this case is a good idea (while still having the known problem section). A more complete solution would be changing the applicability to `MaybeIncorrect` for these scenarios and then using [`multispan_sugg_with_applicability`](https://github.com/rust-lang/rust-clippy/blob/9e53b6544f77de0f4bfb64413014c9ebd832d133/clippy_utils/src/diagnostics.rs#L227)

Reply:
Sorry, do you mean that I should just change `multispan_sugg` to `multispan_sugg_with_applicability` and use `Applicability::MaybeIncorrect`, or check if the `///!` is preceded by `///` and not lint in those cases?

Reply:
It may need to be `MaybeIncorrect` normally, it seems pretty common to accidentally put these comments [after an item](https://github.com/lsd-rs/lsd/blob/118faab5339cdd02ef31a47057173a7cbe91534b/src/config_file.rs#L9-L10) which would be a compile error when fixed